### PR TITLE
add specs with updated implementation

### LINKED ISSUE #80: Avoid HTTP check if version installed matches version pinned in config
My config
```ruby
require "webdrivers"
Webdrivers::Chromedriver.version = '74.0.3729.6'
Webdrivers.logger.level = :DEBUG
```

Output
```sh
➜ bin/rails test:system
Run options: --seed 1780

# Running:

Capybara starting Puma...
* Version 3.12.1 , codename: Llamas in Pajamas
* Min threads: 0, max threads: 4
* Listening on tcp://127.0.0.1:51226
2019-04-29 12:58:13 DEBUG Webdrivers Looking for Site: https://chromedriver.storage.googleapis.com
2019-04-29 12:58:14 DEBUG Webdrivers Get response: #<Net::HTTPOK 200 OK readbody=true>
2019-04-29 12:58:14 DEBUG Webdrivers Found Site: https://chromedriver.storage.googleapis.com
2019-04-29 12:58:14 DEBUG Webdrivers Checking current version
2019-04-29 12:58:14 DEBUG Webdrivers File is already downloaded: true
2019-04-29 12:58:14 DEBUG Webdrivers Current /Users/edgar/.webdrivers/chromedriver version: ChromeDriver 74.0.3729.6 (255758eccf3d244491b8a1317aa76e1ce10d57e9-refs/branch-heads/3729@{#29})

2019-04-29 12:58:14 DEBUG Webdrivers Expected webdriver version found
```

### PR DESCRIPTION
**Pull request by titusfortner**
This would fix #80 

Please make sure that these specs match what you expect. Chrome is a little special, but IE & Firefox are the exact same logic.

~~I updated specs for all of the drivers except for Edge. The existing specs are passing for Edge, but I don't think this code (or existing code works correctly for Edge). I'm going to propose some code changes for that separately before release.~~
I added specs with separate logic for Edge. Things are a little weird with the latest updates: https://developer.microsoft.com/en-us/microsoft-edge/tools/webdriver/
This will need to be updated once we see the pattern that `msedgedriver.exe` takes going forward

1. Adds Custom Errors
2. Minimizes unnecessary network calls
3. Abstracts out a couple methods to make testing easier
4. Chromedriver code is slightly more strict when in offline mode (`#sufficient_binary?`)

Also, I want to release a 3.9 before merging this in.

### CODE REVIEW

File: lib/webdrivers/chromedriver.rb [hunk 1]
```diff
@@ -44,7 +41,7 @@ def latest_point_release(version)
           msg = version > latest_release ? 'you appear to be using a non-production version of Chrome; ' : ''
           msg = "#{msg}please set `Webdrivers::Chromedriver.version = <desired driver version>` to an known "\
```
Comment (twalpole) [inline lib/webdrivers/chromedriver.rb:43]: Needs to be update if swapping to `required_version`

File: lib/webdrivers/mswebdriver.rb [hunk 2]
```diff
@@ -1,38 +1,84 @@
...
 
-      # Webdriver binaries for Microsoft Edge are not backwards compatible.
-      # For this reason, instead of downloading the latest binary we download the correct one for the
-      # currently installed browser version.
-      alias version windows_version
+      def remove
+        @download_url = nil
+        @latest_version = nil
+        Webdrivers.logger.debug 'Deleting MicrosoftWebDriver.exe'
+        FileUtils.rm_f(File.join(install_dir, 'MicrosoftWebDriver.exe'))
```
Comment (twalpole) [inline lib/webdrivers/mswebdriver.rb]: The _f version doesn't return errors if file can't be deleted - Not sure if that's a concern or not

File: spec/webdrivers/chromedriver_spec.rb [hunk 3]
```diff
@@ -5,125 +5,207 @@
...
-  it 'downloads latest release for current version of Chrome by default' do
-    chromedriver.remove
-    chromedriver.download
-    cur_ver    = chromedriver.current_version.version
-    latest_ver = chromedriver.latest_version.version
-    expect(cur_ver).to eq latest_ver
-  end
+      it 'does not download when offline, binary exists and is less than v70' do
+        allow(Net::HTTP).to receive(:get_response).and_raise(SocketError)
+        allow(chromedriver).to receive(:downloaded?).and_return(true)
```
Comment (twalpole) [inline spec/webdrivers/chromedriver_spec.rb]: Not 100% sure but wouldn't this stop it from being downloaded - and not actually test that it wasn't downloaded

File: spec/webdrivers/chromedriver_spec.rb [hunk 4]
```diff
@@ -5,125 +5,207 @@
...
-  end
+        expect(File.exist?(chromedriver.binary)).to be false
+      end
 
-  it 'removes chromedriver' do
-    chromedriver.remove
-    expect(chromedriver.current_version).to be_nil
+      it 'does not download when offline, binary exists and matches major browser version' do
+        allow(Net::HTTP).to receive(:get_response).and_raise(SocketError)
+        allow(chromedriver).to receive(:downloaded?).and_return(true)
```
Comment (twalpole) [inline spec/webdrivers/chromedriver_spec.rb]: Same concern as previous - again not sure I'm correct on this though

File: spec/webdrivers/chromedriver_spec.rb [hunk 5]
```diff
@@ -5,125 +5,207 @@
...
+        allow(chromedriver).to receive(:current_version).and_return(Gem::Version.new('73.0.3683.20'))
+
+        chromedriver.update
+
+        expect(File.exist?(chromedriver.binary)).to be false
+      end
+
+      it 'raises ConnectionError when offline, and binary does not match major browser version' do
+        allow(Net::HTTP).to receive(:get_response).and_raise(SocketError)
+        allow(chromedriver).to receive(:downloaded?).and_return(true)
```
Comment (twalpole) [inline spec/webdrivers/chromedriver_spec.rb]: Same

File: spec/webdrivers/chromedriver_spec.rb [hunk 6]
```diff
@@ -5,125 +5,207 @@
...
+      allow(chromedriver).to receive(:chrome_version).and_return '71.0.3578.9999'
+
+      expect(chromedriver.latest_version).to eq Gem::Version.new('71.0.3578.137')
     end
 
-    it 'raises an exception for beta version' do
+    it 'raises VersionError for beta version' do
       allow(chromedriver).to receive(:chrome_version).and_return('100.0.0')
       msg = 'you appear to be using a non-production version of Chrome; please set '\
 '`Webdrivers::Chromedriver.version = <desired driver version>` to an known chromedriver version: '\
```
Comment (twalpole) [inline spec/webdrivers/chromedriver_spec.rb:131]: `required_version` if changing

**Review by twalpole — COMMENTED** [review]
After a quick read through I made a couple of comments.  The biggest thing I see is the use of `allow(...).to receive(:downloaded?).and_return(true)` in the tests that are verifying downloads don't happen.  I'm not 100% correct it's wrong but the naming seems like it could be.  I'll have to read more through the code to see exactly what happens with `downloaded?`

File: spec/webdrivers/chromedriver_spec.rb (hunk 3 above)
Comment (titusfortner) [reply spec/webdrivers/chromedriver_spec.rb]: This is testing if it is offline, and a binary is there, it doesn't still download.
setting `#downloaded?` to true is mocking that something is there. Then if the file doesn't actually exist, then we know it wasn't still downloaded. It's just stepping through the conditional in the method to make sure it short-circuits.

File: lib/webdrivers/chromedriver.rb (hunk 1 above)
Comment (titusfortner) [reply lib/webdrivers/chromedriver.rb:43]: I was going to address everything from #36 is a separate PR

File: lib/webdrivers/mswebdriver.rb (hunk 2 above)
Comment (kapoorlakshya) [reply lib/webdrivers/mswebdriver.rb]: Yeah, same as #87.

Comment (titusfortner) [reply lib/webdrivers/mswebdriver.rb]: Removed this code

File: lib/webdrivers/common.rb [hunk 7]
```diff
@@ -9,45 +9,21 @@ class << self
...
-        # Can't find desired and no existing binary
-        if desired_version.nil?
-          msg = "Unable to find the latest version of #{file_name}; try downloading manually " \
-"from #{base_url} and place in #{install_dir}"
-          raise StandardError, msg
-        end
-
         if correct_binary?
-          Webdrivers.logger.debug 'Expected webdriver version found'
+          Webdrivers.logger.debug 'The desired webdriver version is already on the system'
```
Comment (kapoorlakshya) [inline lib/webdrivers/common.rb]: Maybe replace "desired" with "required" to match `required_version`?

File: lib/webdrivers/mswebdriver.rb [hunk 8]
```diff
@@ -1,32 +1,46 @@
 # frozen_string_literal: true
 
+require 'nokogiri'
+
 module Webdrivers
   class MSWebdriver < Common
     class << self
+      attr_accessor :ignore
```
Comment (kapoorlakshya) [inline lib/webdrivers/mswebdriver.rb]: What's the purpose of this?

Comment (titusfortner) [reply lib/webdrivers/mswebdriver.rb]: I'm not sure MS webdriver works with webdrivers. This is a toggle for turning off the gem functionality just for Edge.

Comment (kapoorlakshya) [reply lib/webdrivers/mswebdriver.rb]: I was not sure of the use case, but I get it now after looking at #102.

**Review by kapoorlakshya — APPROVED** [review]
Let's merge once specs pass.

**Pull request merged**